### PR TITLE
'Large Party Map' Performance - Part 2/x

### PR DESCRIPTION
**Pull request by 0xdevalias**
Part 1: https://github.com/sparkletown/sparkle/pull/726 (though this PR isn't directly dependant on it)

This PR contains a bunch of smaller cleanup parts as part of improving the performance of `src/components/templates/Camp/components/Map.tsx`. I decided to split them out into this PR to minimize the complexity of the bigger changes to come.

Main changes:

- 5447ea43 add useCampPartygoers() hook + minor cleanup in useConnectPartyGoers()
- 5eb19306 refactor Camp to use useCampPartygoers() hook + static selector reference for campVenueSelector
- refactor isExternalUrl, getRoomUrl, openRoomUrl, openUrl into utils + updates references
- add currentVenueSelectorData, orderedVenuesSelector + refactor components to use them + some minor cleanup on the way

See [commits](https://github.com/sparkletown/sparkle/pull/727/commits) for the rest.

### CODE REVIEW

**Comment by 0xdevalias** [comment]
Rebased onto current `staging` branch to simplify some of my changes with a cleaner history.

File: src/components/templates/Playa/Playa.tsx [hunk 1]
```diff
@@ -6,15 +6,13 @@ import React, {
   useMemo,
```
Comment (0xdevalias) [inline src/components/templates/Playa/Playa.tsx:6]: [FYI] most of the import changes here are just due to it being sorted, main changes was the selectors + related imports

File: src/components/templates/Camp/components/Map.tsx [hunk 2]
```diff
@@ -174,13 +185,6 @@ export const Map: React.FC<PropsType> = ({
     [columns, isHittingRoom, rooms, rows, selectedRoom, setSelectedRoom]
   );
 
-  const isExternalLink = useCallback(
-    (url: string) =>
-      url.includes("http") &&
-      new URL(window.location.href).host !== new URL(getRoomUrl(url)).host,
-    []
-  );
```
Comment (0xdevalias) [inline src/components/templates/Camp/components/Map.tsx:182]: [FYI] this was refactored into `utils/url`

File: src/components/templates/Camp/components/Map.tsx [hunk 3]
```diff
@@ -236,12 +240,7 @@ export const Map: React.FC<PropsType> = ({
       if (enterPress && selectedRoom) {
         setKeyDown(true);
         setTimeout(() => setKeyDown(false), MOVEMENT_INTERVAL);
-
-        const isExternalUrl = isExternalLink(selectedRoom.url);
-        window.open(
-          getRoomUrl(selectedRoom.url),
-          isExternalUrl ? "_blank" : "noopener,noreferrer"
-        );
```
Comment (0xdevalias) [inline src/components/templates/Camp/components/Map.tsx:244]: [FYI] this was refactored into `utils/url`

File: src/components/templates/Camp/components/Map.tsx [hunk 4]
```diff
@@ -315,10 +313,6 @@ export const Map: React.FC<PropsType> = ({
     }
   }
 
-  const getRoomUrl = (roomUrl: string) => {
-    return roomUrl.includes("http") ? roomUrl : "//" + roomUrl;
-  };
```
Comment (0xdevalias) [inline src/components/templates/Camp/components/Map.tsx:320]: [FYI] this was refactored into `utils/url`

File: src/components/templates/Camp/Camp.tsx [hunk 5]
```diff
@@ -1,54 +1,43 @@
...
-  }, [setNowMs]);
-
-  const usersInCamp = partygoers
-    ? partygoers.filter(
-        (partygoer) =>
-          partygoer?.lastSeenIn &&
-          partygoer?.lastSeenIn[venue.name] >
-            (nowMs - LOC_UPDATE_FREQ_MS * 2) / 1000
-      )
-    : [];
```
Comment (0xdevalias) [inline src/components/templates/Camp/Camp.tsx:51]: [FYI] this was refactored into `src/hooks/useCampPartygoers.ts`

File: src/components/templates/Camp/Camp.tsx [hunk 6]
```diff
@@ -1,54 +1,43 @@
...
+
+const campVenueSelector = (state: RootState) =>
+  state.firestore.ordered.currentVenue?.[0] as CampVenue;
 
 const Camp: React.FC = () => {
-  useConnectPartyGoers();
   const [isRoomModalOpen, setIsRoomModalOpen] = useState(false);
   const [selectedRoom, setSelectedRoom] = useState<CampRoomData | undefined>();
   const [showEventSchedule, setShowEventSchedule] = useState(false);
-  const [nowMs, setNowMs] = useState(new Date().getTime());
```
Comment (0xdevalias) [inline src/components/templates/Camp/Camp.tsx:29]: [FYI] this was refactored into `src/hooks/useCampPartygoers.ts`

File: src/components/templates/Camp/Camp.tsx [hunk 7]
```diff
@@ -1,54 +1,43 @@
...
 
-  const { partygoers, venue } = useSelector((state) => ({
-    venue: state.firestore.ordered.currentVenue?.[0] as CampVenue,
-    partygoers: state.firestore.ordered.partygoers,
```
Comment (0xdevalias) [inline src/components/templates/Camp/Camp.tsx:33]: [FYI] this was refactored into `src/hooks/useCampPartygoers.ts`

**Review by 0xdevalias — COMMENTED** [review]
Added a few FYI comments to help with reviewing, most of the files changed are only a couple of lines each (due to selector fixes), shouldn't take too long to skim through those bits.

File: src/hooks/useConnectPartyGoers.ts [hunk 8]
```diff
@@ -5,6 +5,7 @@ import { partygoersSelector } from "utils/selectors";
 import { useSelector } from "./useSelector";
 import { useUserLastSeenLimit } from "./useUserLastSeenLimit";
 
+// @debt rename this useConnectPartygoers for consistency
```
Comment (thecadams) [inline src/hooks/useConnectPartyGoers.ts:8]: +1

**Review by thecadams — APPROVED** [review]
lgtm! Going to merge this as part of today's drive to get all the merging done- hope that's OK.

File: src/components/templates/Camp/components/RoomModal.tsx [hunk 9]
```diff
@@ -1,39 +1,50 @@
...
 
-import "../../../templates/PartyMap/RoomModal/RoomModal.scss";
-import { currentTimeInUnixEpoch, ONE_MINUTE_IN_SECONDS } from "utils/time";
-import { useFirestoreConnect } from "react-redux-firebase";
+import UserList from "components/molecules/UserList";
+
+import { RoomModalOngoingEvent } from "./RoomModalOngoingEvent";
+import { ScheduleItem } from "./ScheduleItem";
+
+import "./RoomModal.scss";
```
Comment (0xdevalias) [inline src/components/templates/Camp/components/RoomModal.tsx:24]:
```suggestion
import "templates/PartyMap/RoomModal/RoomModal.scss";
```

File: src/components/templates/Camp/components/RoomModalOngoingEvent.tsx [hunk 10]
```diff
@@ -1,39 +1,36 @@
...
 
-import "../../../templates/PartyMap/components/RoomModalOngoingEvent/RoomModalOngoingEvent.scss";
+import { getCurrentEvent } from "utils/event";
+import { isExternalUrl } from "utils/url";
+
 import { useDispatch } from "hooks/useDispatch";
-import { retainAttendance } from "store/actions/Attendance";
 
-interface PropsType {
+import "./RoomModalOngoingEvent.scss";
```
Comment (0xdevalias) [inline src/components/templates/Camp/components/RoomModalOngoingEvent.tsx:13]:
```suggestion
import "templates/PartyMap/components/RoomModalOngoingEvent/RoomModalOngoingEvent.scss";
```

**Comment by 0xdevalias** [comment]
> @thecadams Going to merge this as part of today's drive to get all the merging done- hope that's OK.

Happy for you to once you fix those broken bits you noticed that I missed :p <3

**Pull request merged**